compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java

```diff
@@ -283,9 +283,7 @@ BLangStatementExpression desugar(BLangQueryExpr queryExpr, SymbolEnv env,
             onConflictExpr = (onConflictExpr == null)
                     ? ASTBuilderUtil.createLiteral(pos, symTable.nilType, Names.NIL_VALUE)
                     : onConflictExpr;
-            BMapType mapType = (BMapType) ((BUnionType) queryExpr.getBType()).getMemberTypes()
-                    .stream().filter(m -> Types.getReferredType(m).tag == TypeTags.MAP)
-                    .findFirst().orElse(symTable.mapType);
+            BMapType mapType = (BMapType) getMapType(queryExpr.getBType());
```

Comment:
Why wouldn't the queryExpr type be resolved to a single type? Ideally, it should.

Reply:
This happens in 2 scenarios. 
1. When the contextually expected type is readonly, type of the query expression will be a `BIntersectionType` type.
2. When `On-Conflict` clause is available, type of the query expression will be a `BUnionType` of `BMapType` and `BError` type.

Reply:
That makes sense